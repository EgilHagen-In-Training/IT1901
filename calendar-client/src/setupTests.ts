// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import { User } from "./types/user";

export const mockUser: User = {
	username: "testUser",
	calendar: {
		events: [
			{
				header: "testEvent1",
				description: "testDescription1",
				date: "2021-11-23",
				timeString: "12:30",
				timeHour: 12,
				timeMinute: 30,
				id: 1,
			},
			{
				header: "testEvent2",
				description: "testDescription1",
				date: "2021-11-11",
				timeString: "15:00",
				timeHour: 15,
				timeMinute: 0,
				id: 2,
			},
		],
	},
};
