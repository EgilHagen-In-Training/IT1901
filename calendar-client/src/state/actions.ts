import { User } from "../types/user";

export enum ActionTypes {
  SET_USER = "SET_USER",
  LOGOUT_USER = "LOGOUT_USER",
}

type SetUserAction = { type: ActionTypes.SET_USER; payload: User };
type LogoutUserAction = { type: ActionTypes.LOGOUT_USER };


export type Action =
  | SetUserAction
  | LogoutUserAction

export const setUser = (user: User): SetUserAction => ({
	type: ActionTypes.SET_USER,
	payload: user,
});

export const logoutUser = (): LogoutUserAction => ({
	type: ActionTypes.LOGOUT_USER,
});

