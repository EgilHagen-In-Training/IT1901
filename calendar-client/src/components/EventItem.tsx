import {
	Box,
	IconButton,
	HStack,
	Popover,
	PopoverTrigger,
	PopoverContent,
	PopoverArrow,
	PopoverCloseButton,
	PopoverHeader,
	PopoverBody,
	Button,
	Center,
} from "@chakra-ui/react";
import React, { FC, useContext } from "react";
import { EventType } from "../types/user";
import { useHistory } from "react-router-dom";
import { StateContext } from "../state/state";
import { setEvent, setUser } from "../state/actions";
import { DeleteIcon } from "@chakra-ui/icons";
import { deleteEvent, getUser } from "../api";
type EventItemProps = {
	event: EventType;
};
const EventItem: FC<EventItemProps> = ({ event }: EventItemProps) => {
	const history = useHistory();
	const { state, dispatch } = useContext(StateContext);
	const { user } = state;
	const editEvent = () => {
		dispatch(setEvent(event));
		history.push("/event");
	};

	if (!user) return <div>Loading...</div>;

	const onDelete = async () => {
		await deleteEvent(user.username, event.id);
		const reqUser = await getUser(user.username);
		if (reqUser) {
			dispatch(setUser(reqUser));
		}
	};

	return (
		<HStack
			as="button"
			textAlign="left"
			w="100%"
			_hover={{ background: "yellow.500" }}
			transition="0.2s"
		>
			<Box onClick={() => editEvent()}>
				{event.timeString} {event.header}
			</Box>
			<Popover id="pop-id">
				<PopoverTrigger>
					<IconButton
						aria-label="delete"
						icon={<DeleteIcon />}
						size="sm"
						colorScheme="tomato"
						_hover={{ color: "black" }}
					/>
				</PopoverTrigger>
				<PopoverContent>
					<PopoverArrow />
					<PopoverCloseButton />
					<PopoverHeader>Delete event</PopoverHeader>
					<PopoverBody>
						{"Are you sure you want to delete that event?"}
						<Center>
							<Button onClick={() => onDelete()}>Delete</Button>
						</Center>
					</PopoverBody>
				</PopoverContent>
			</Popover>
		</HStack>
	);
};

export default EventItem;
