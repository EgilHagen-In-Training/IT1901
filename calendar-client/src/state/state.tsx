import React, { createContext, Dispatch, ReactNode, useReducer } from "react";
import { User, EventType } from "../types/user";
import { Action, ActionTypes } from "./actions";

type State = {
	user?: User;
	currentEvent?: EventType;
};

const initialState: State = {
	user: undefined,
	currentEvent: undefined,
};

const StateReducer = (state: State, action: Action): State => {
	switch (action.type) {
		case ActionTypes.SET_USER:
			return { ...state, user: action.payload };
		case ActionTypes.LOGOUT_USER:
			return {
				...state,
				user: undefined,
			};
		case ActionTypes.SET_EVENT:
			return { ...state, currentEvent: action.payload };
		case ActionTypes.CLEAR_EVENT:
			return { ...state, currentEvent: undefined };
		default:
			return { ...state };
	}
};

type StateContextProps = {
	state: State;
	dispatch: Dispatch<Action>;
};

const initialContext: StateContextProps = {
	state: initialState,
	dispatch: () => null,
};

export const StateContext = createContext<StateContextProps>(initialContext);

type StateProviderProps = {
	children: ReactNode;
};

export const StateProvider = (props: StateProviderProps): JSX.Element => {
	const [state, dispatch] = useReducer(StateReducer, initialState);
	const { children } = props;

	return (
		<StateContext.Provider value={{ state, dispatch }}>
			{children}
		</StateContext.Provider>
	);
};
