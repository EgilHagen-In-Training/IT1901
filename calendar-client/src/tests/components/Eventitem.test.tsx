import React from "react";
import EventItem from "../../../src/components/EventItem";
import renderer from "react-test-renderer";
import { StateContext } from "../../state/state";
import { mockUser, MockEvent } from "../../setupTests";

test("EventItem renders correctly", () => {
	const state = {
		user: mockUser,
	};
	const dispatch = jest.fn();

	const tree = renderer
		.create(
			<StateContext.Provider value={{ state, dispatch }}>
				<EventItem event={MockEvent} />
			</StateContext.Provider>
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
