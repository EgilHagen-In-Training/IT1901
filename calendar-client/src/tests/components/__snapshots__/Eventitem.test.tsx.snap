// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventItem renders correctly 1`] = `
<button
  className="chakra-stack css-1pq0kbc"
>
  <div
    className="css-0"
    onClick={[Function]}
  >
    12:30
     
    testEvent1
  </div>
  <button
    aria-controls="popover-content-pop-id"
    aria-expanded={false}
    aria-haspopup="dialog"
    aria-label="delete"
    className="chakra-button css-ve4un"
    id="popover-trigger-pop-id"
    onClick={[Function]}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="chakra-icon css-onkibi"
      focusable={false}
      viewBox="0 0 24 24"
    >
      <g
        fill="currentColor"
      >
        <path
          d="M19.452 7.5H4.547a.5.5 0 00-.5.545l1.287 14.136A2 2 0 007.326 24h9.347a2 2 0 001.992-1.819L19.95 8.045a.5.5 0 00-.129-.382.5.5 0 00-.369-.163zm-9.2 13a.75.75 0 01-1.5 0v-9a.75.75 0 011.5 0zm5 0a.75.75 0 01-1.5 0v-9a.75.75 0 011.5 0zM22 4h-4.75a.25.25 0 01-.25-.25V2.5A2.5 2.5 0 0014.5 0h-5A2.5 2.5 0 007 2.5v1.25a.25.25 0 01-.25.25H2a1 1 0 000 2h20a1 1 0 000-2zM9 3.75V2.5a.5.5 0 01.5-.5h5a.5.5 0 01.5.5v1.25a.25.25 0 01-.25.25h-5.5A.25.25 0 019 3.75z"
        />
      </g>
    </svg>
  </button>
  <div
    className="chakra-popover__popper css-0"
    style={
      Object {
        "inset": "0 auto auto 0",
        "minWidth": "max-content",
        "position": "absolute",
        "visibility": "hidden",
      }
    }
  >
    <section
      className="chakra-popover__content css-1ab1jlb"
      id="popover-content-pop-id"
      onBlur={[Function]}
      onKeyDown={[Function]}
      role="dialog"
      style={
        Object {
          "--popper-arrow-shadow-color": undefined,
          "--popper-arrow-size": undefined,
          "opacity": 0,
          "transform": "scale(0.95) translateZ(0)",
          "transformOrigin": "var(--popper-transform-origin)",
          "visibility": "hidden",
        }
      }
      tabIndex={-1}
    >
      <div
        className="chakra-popover__arrow-positioner css-0"
        data-popper-arrow=""
        style={
          Object {
            "position": "absolute",
          }
        }
      >
        <div
          className="chakra-popover__arrow css-0"
          data-popper-arrow-inner=""
        />
      </div>
      <button
        aria-label="Close"
        className="chakra-popover__close-btn css-6sdaxy"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <header
        className="chakra-popover__header css-0"
        id="popover-header-pop-id"
      >
        Delete event
      </header>
      <div
        className="chakra-popover__body css-0"
        id="popover-body-pop-id"
      >
        Are you sure you want to delete that event?
        <div
          className="css-gmuwbf"
        >
          <button
            className="chakra-button css-5c2176"
            onClick={[Function]}
            type="button"
          >
            Delete
          </button>
        </div>
      </div>
    </section>
  </div>
</button>
`;
